### ISSUE TITLE
False negative on RET503 when using `match`

### ISSUE THREAD

**Issue by lihu-zhong** [astral-sh/ruff#23206]
### Summary

Given the following code, RET503 only warns on the first function:

```python
def ant(arg: bool) -> str | None:
    if arg:
        return "bat"


def cat(arg: bool) -> str | None:
    match arg:
        case True:
            return "dog"
```

To me, the structure and reasoning for warning in these two patterns is identical, and I'd expect a warning that I have an implicit return value. That said, I could imagine fixing this might make https://github.com/astral-sh/ruff/issues/18656 more annoying, so maybe it's intentional? Thought I'd report it either way.

I have linked the reproducing playground:
https://play.ruff.rs/9d21e1a1-cec4-4c06-93ba-f8705a816730

### Version

ruff 0.14.10

**Comment by ntBre** [comment]
Interesting, it does seem like we could do something better here, even without type inference. Our `match` analysis here seems quite a bit more limited than  what we do for `if` statements earlier in the same function:

https://github.com/astral-sh/ruff/blob/e6dc572bb0848946051c7482fc4103070e309369/crates/ruff_linter/src/rules/flake8_return/rules/function.rs#L526-L530

Thanks for the report!

**Comment by ntBre** [comment]
I just commented on the linked PR, but this may be trickier than I thought initially since `match` statements don't have to be exhaustive. One example from the ecosystem report is much less clear than the example here where it's "obvious" that the `match` on a `bool` is non-exhaustive:

```py
    def _get_target_temperature(self) -> float | None:
        """Return the target temperature."""


        match self._eq3_config.target_temp_selector:
            case TargetTemperatureSelector.TARGET:
                return self._target_temperature
            case TargetTemperatureSelector.LAST_REPORTED:
                return float(self._thermostat.status.target_temperature)
```

I think we may need to wait on better type inference to improve this.
